src/librbd/BlockGuard.h

```diff
@@ -81,8 +81,18 @@ class BlockGuard {
       // request against an already detained block
       detained_block_extent = &(*it);
       if (block_operation != nullptr) {
+        m_detained_block_extents.erase(it);
         detained_block_extent->block_operations.emplace_back(
           std::move(*block_operation));
+
+        if (block_extent.block_start < detained_block_extent->block_extent.block_start) {
+          detained_block_extent->block_extent.block_start = block_extent.block_start;
+        }
+        if (block_extent.block_end > detained_block_extent->block_extent.block_end) {
+          detained_block_extent->block_extent.block_end = block_extent.block_end;
+        }
+
+        m_detained_block_extents.insert(*detained_block_extent);
```

Comment:
This seems broken to me.  Consider:
1. detain(3, 5)
       inserts (3, 5)
2. detain(8, 11)
       inserts (8, 11)
3. detain(4, 9)
       finds (3, 5), removes it
       adjusts (3, 5) to (3, 9), attempts to insert it

But (3, 9) can't be inserted into the set because it overlaps with (8, 11).  The entry is lost...

Reply:
Yes, you are right.　The situation is more complicated than I thought．for a(3,5), b(4, 9), c(8, 11), it depend on insert order .